File: internal/devenv/dockercli/engine.go
package dockercli

import (
	"encoding/json"
	"os/exec"
)

type Engine string

const (
	EngineDocker Engine = "docker"
	EnginePodman Engine = "podman"
)

type EngineInfo struct {
	Engine        Engine
	ServerVersion string
	ComposePlugin bool
	SocketPath    string
	Rootless      bool
}

type InfoRunner func(dockerBin string, args ...string) ([]byte, error)

func dockerFallbackEngineInfo(socketPath string) EngineInfo {
	return EngineInfo{Engine: EngineDocker, ServerVersion: "unknown", SocketPath: socketPath}
}

type dockerInfoShape struct {
	ServerVersion string `json:"ServerVersion"`
	ClientInfo    struct {
		Plugins []struct {
			Name string `json:"Name"`
		} `json:"Plugins"`
	} `json:"ClientInfo"`
}

type podmanInfoShape struct {
	Version struct {
		Version string `json:"Version"`
	} `json:"version"`
	Host struct {
		RemoteSocket struct {
			Path string `json:"path"`
		} `json:"remoteSocket"`
		Security struct {
			Rootless bool `json:"rootless"`
		} `json:"security"`
	} `json:"host"`
	Store struct {
		GraphDriverName string `json:"graphDriverName"`
	} `json:"store"`
}

func hasComposePlugin(shape dockerInfoShape) bool {
	for _, p := range shape.ClientInfo.Plugins {
		if p.Name == "compose" {
			return true
		}
	}
	return false
}

func parseDockerInfo(raw []byte, socketPath string) (EngineInfo, bool) {
	var shape dockerInfoShape
	if err := json.Unmarshal(raw, &shape); err != nil || shape.ServerVersion == "" {
		return EngineInfo{}, false
	}
	return EngineInfo{
		Engine:        EngineDocker,
		ServerVersion: shape.ServerVersion,
		ComposePlugin: hasComposePlugin(shape),
		SocketPath:    socketPath,
	}, true
}

func parsePodmanInfo(raw []byte, socketPath string) (EngineInfo, bool) {
	var shape podmanInfoShape
	if err := json.Unmarshal(raw, &shape); err != nil {
		return EngineInfo{}, false
	}
	if shape.Version.Version == "" || (shape.Host.RemoteSocket.Path == "" && shape.Store.GraphDriverName == "") {
		return EngineInfo{}, false
	}
	resolvedSocket := socketPath
	if resolvedSocket == "" {
		resolvedSocket = shape.Host.RemoteSocket.Path
	}
	return EngineInfo{
		Engine:        EnginePodman,
		ServerVersion: shape.Version.Version,
		SocketPath:    resolvedSocket,
		Rootless:      shape.Host.Security.Rootless,
	}, true
}

func DetectEngine(dockerBin, socketPath string, run InfoRunner) (EngineInfo, error) {
	raw, err := run(dockerBin, "info", "--format", "json")
	if err != nil {
		return EngineInfo{}, err
	}
	if info, ok := parseDockerInfo(raw, socketPath); ok {
		return info, nil
	}
	if info, ok := parsePodmanInfo(raw, socketPath); ok {
		return info, nil
	}
	return dockerFallbackEngineInfo(socketPath), nil
}

func (r *Runner) DetectEngine(socketPath string) (EngineInfo, error) {
	return DetectEngine(r.dockerBin(), socketPath, func(dockerBin string, args ...string) ([]byte, error) {
		return exec.Command(dockerBin, args...).Output()
	})
}

File: src/lib/dev-environment/engine.ts
import { execFile } from 'node:child_process';
import { promisify } from 'node:util';

const execFileAsync = promisify( execFile );

export type ContainerEngine = 'docker' | 'podman';

export interface EngineInfo {
	engine: ContainerEngine;
	serverVersion: string;
	composePlugin?: string;
	socketPath: string;
	rootless: boolean;
}

type ExecFn = typeof execFileAsync;

const fallbackEngineInfo = ( socketPath: string ): EngineInfo => ( {
	engine: 'docker',
	serverVersion: 'unknown',
	socketPath,
	rootless: false,
} );

const isRecord = ( value: unknown ): value is Record< string, unknown > =>
	typeof value === 'object' && value !== null;

const isDockerShape = (
	info: Record< string, unknown >
): info is Record< string, unknown > & { ServerVersion: string } =>
	typeof info.ServerVersion === 'string';

const isPodmanShape = ( info: Record< string, unknown > ): boolean => {
	const version = info.version;
	const host = info.host;
	const store = info.store;

	const hasVersion = isRecord( version ) && typeof version.Version === 'string';
	const hasPodmanHostShape = isRecord( host ) && 'remoteSocket' in host;
	const hasPodmanStoreShape = isRecord( store ) && typeof store.graphDriverName === 'string';

	return hasVersion && ( hasPodmanHostShape || hasPodmanStoreShape );
};

const findComposePluginVersion = ( info: Record< string, unknown > ): string | undefined => {
	const clientInfo = info.ClientInfo;
	if ( ! isRecord( clientInfo ) ) {
		return undefined;
	}

	const plugins: unknown = clientInfo.Plugins;
	if ( ! Array.isArray( plugins ) ) {
		return undefined;
	}

	const composePlugin: unknown = ( plugins as unknown[] ).find(
		plugin => isRecord( plugin ) && plugin.Name === 'compose'
	);

	return isRecord( composePlugin ) && typeof composePlugin.Version === 'string'
		? composePlugin.Version
		: undefined;
};

const isPodmanRootless = ( info: Record< string, unknown > ): boolean => {
	const host = info.host;
	if ( ! isRecord( host ) ) {
		return false;
	}

	const security = host.security;
	return isRecord( security ) && security.rootless === true;
};

const parseDockerInfo = ( info: Record< string, unknown >, socketPath: string ): EngineInfo => ( {
	engine: 'docker',
	serverVersion: info.ServerVersion as string,
	composePlugin: findComposePluginVersion( info ),
	socketPath,
	rootless: false,
} );

const parsePodmanInfo = ( info: Record< string, unknown >, socketPath: string ): EngineInfo => {
	const version = info.version as Record< string, unknown >;
	return {
		engine: 'podman',
		serverVersion: version.Version as string,
		composePlugin: findComposePluginVersion( info ),
		socketPath,
		rootless: isPodmanRootless( info ),
	};
};

export async function detectEngine(
	dockerBin: string,
	socketPath: string,
	exec: ExecFn = execFileAsync
): Promise< EngineInfo > {
	try {
		const { stdout } = await exec( dockerBin, [ 'info', '--format', 'json' ] );
		const info: unknown = JSON.parse( String( stdout ) );

		if ( ! isRecord( info ) ) {
			return fallbackEngineInfo( socketPath );
		}

		if ( isDockerShape( info ) ) {
			return parseDockerInfo( info, socketPath );
		}

		if ( isPodmanShape( info ) ) {
			return parsePodmanInfo( info, socketPath );
		}

		return fallbackEngineInfo( socketPath );
	} catch {
		return fallbackEngineInfo( socketPath );
	}
}
